util/licence-checker.py

```diff
@@ -0,0 +1,357 @@
+        self.comment_prefix = str(prefix)
+        self.comment_suffix = str(suffix)
+        self.first_line_prefix = self.comment_prefix
+
+    def search_line(self, licence_first_word):
+        return self.comment_prefix + ' ' + licence_first_word
+
+    def expected_full_line(self, licence_line):
+        return self.comment_prefix + ' ' + licence_line + ' ' + self.comment_suffix
+
+
+SLASH_SLASH = LineCommentStyle("//")
+HASH = LineCommentStyle("#")
+
+# (Priortised) Mapping of file name suffixes to CommentStyle object.
+# If the suffix of your file does not match one of these, it will not be
+# checked.
+COMMENT_CHARS = [
+    # Hardware Files
+    ([".svh", ".sv", ".sv.tpl"], SLASH_SLASH),  # SystemVerilog
+    # Hardware Build Systems
+    ([".tcl"], HASH),  # tcl
+    ([".core", ".core.tpl"], DifferentFirstLineCommentStyle("#", "CAPI=2")),  # FuseSoC Core Files
+    (["Makefile", ".mk"], HASH),  # Makefiles
+    ([".ys"], HASH),  # Yosys script
+    ([".waiver"], HASH),  # Wavier files
+    ([".vlt"], SLASH_SLASH),  # More wavier files
+    ([".vbl"], HASH),  # Verible configuration
+    ([".el", ".el.tpl"], SLASH_SLASH), # Exclusion List
+    ([".do"], SLASH_SLASH),
+    ([".sdc"], HASH),
+    ([".f"], None),  # Dependency Files
+    # Software Files
+    ([".c", ".c.tpl", ".h", ".h.tpl", ".cc", ".cpp"], SLASH_SLASH),  # C,C++
+    ([".S"], SLASH_SLASH),  # Assembly
+    ([".ld"], BlockCommentStyle("/*", "*/")),  # Linker Scripts
+    ([".rs"], SLASH_SLASH),  # Rust
+    # Software Build Systems
+    (["meson.build", "toolchain.txt", "meson_options.txt"], HASH),  # Meson
+    # General Tooling
+    ([".py"], HASH),  # Python
+    ([".sh"], HASH),  # Shell Scripts
+    (["Dockerfile"], HASH),  # Dockerfiles
+    # Configuration
+    ([".hjson", ".hjson.tpl"], SLASH_SLASH),  # hjson
+    ([".yml", ".yaml"], HASH),  # YAML
+    ([".toml"], HASH),  # TOML
+    (["-requirements.txt"], HASH),  # Apt + Python Requirements
+    (["redirector.conf"], HASH),  # nginx config
+    # Documentation
+    ([".md", ".md.tpl", ".html"], None),  # Markdown and HTML
+    ([".css"], BlockCommentStyle("/*", "*/")),  # CSS
+    ([".scss"], SLASH_SLASH),  # SCSS
+    # Unsupported: Explicitly Skip These
+    ([".png", ".svg", ".ico"], None), # Image formats
+    ([], None),
+]
+
+
+def detect_comment_char(filename):
+    for (suffixes, commentstyle) in COMMENT_CHARS:
+        for suffix in suffixes:
+            if filename.endswith(suffix):
+                return commentstyle
+
+    return None
+
+
+def git_find_repo_toplevel():
+    git_output = subprocess.check_output(
+        ['git', 'rev-parse', '--show-toplevel'])
+    return Path(git_output.decode().strip()).resolve()
+
+
+def git_find_all_file_paths(top_level, search_paths):
+    git_output = subprocess.check_output(
+        ["git", "-C", str(top_level), "ls-files", "-z", "--", *search_paths])
+    git_output = git_output.rstrip(b"\0")
+    for path in git_output.split(b"\0"):
```

Comment:
Nit: I'd probably fold the rstrip and split together, so `for path in git_output.rstrip(b'\0').split(b'\0')` to keep things "more SSA". But I don't think it really matters.